apps/src/p5lab/spritelab/libraries/PoemBotLibrary.js

```diff
@@ -188,47 +179,167 @@ export default class PoemBotLibrary extends CoreLibrary {
     return scaledSize;
   }
 
-  drawPoem() {
+  applyEffect(renderInfo, effect, frameCount) {
+    const newLines = [];
+    renderInfo.lines.forEach(line => {
+      const newLine = {...line};
+      if (frameCount >= newLine.start && frameCount < newLine.end) {
+        const progress =
+          (frameCount - newLine.start) / (newLine.end - newLine.start);
+        switch (effect.name) {
+          case 'fade':
+            newLine.alpha = progress * 255;
+            break;
+          case 'typewriter': {
+            const numCharsToShow = Math.floor(progress * newLine.text.length);
+            newLine.text = newLine.text.substring(0, numCharsToShow);
+            break;
+          }
+          case 'flyLeft': {
+            const start = -PLAYSPACE_SIZE / 2;
+            const end = newLine.x;
+            newLine.x = start - progress * (start - end);
+            break;
+          }
+          case 'flyRight': {
+            const start = PLAYSPACE_SIZE * 1.5;
+            const end = newLine.x;
+            newLine.x = start - progress * (start - end);
+            break;
+          }
+          case 'flyTop': {
+            const start = -LINE_HEIGHT;
+            const end = newLine.y;
+            newLine.y = start - progress * (start - end);
+            break;
+          }
+          case 'flyBottom': {
+            const start = PLAYSPACE_SIZE + LINE_HEIGHT;
+            const end = newLine.y;
+            newLine.y = start - progress * (start - end);
+            break;
+          }
+          default:
+            break;
+        }
+      }
+      newLines.push(newLine);
+    });
+    return {
+      ...renderInfo,
+      lines: newLines
+    };
+  }
+
+  applyGlobalLineAnimation(renderInfo, frameCount) {
+    const progress = frameCount / POEM_DURATION;
+    const framesPerLine = POEM_DURATION / renderInfo.lines.length;
+    const newLines = [];
+    for (let i = 0; i < renderInfo.lines.length; i++) {
+      const newLine = {...renderInfo.lines[i]};
+      newLine.start = i * framesPerLine;
+      newLine.end = (i + 1) * framesPerLine;
+      newLines.push(newLine);
+    }
+
+    const numLinesToShow = Math.floor(progress * renderInfo.lines.length);
+    return {
+      ...renderInfo,
+      lines: newLines.slice(0, numLinesToShow + 1) // end index is not inclusive, so + 1
+    };
+  }
+
+  getRenderInfo(poemState, frameCount) {
+    if (!poemState.isVisible) {
+      return {
+        lines: []
+      };
+    }
     let yCursor = OUTER_MARGIN;
-    this.p5.textSize(FONT_SIZE);
-    this.p5.textAlign(this.p5.CENTER);
-    if (this.poem.title) {
-      this.p5.textSize(this.getScaledFontSize(this.poem.title, FONT_SIZE * 2));
-      this.p5.text(this.poem.title, PLAYSPACE_SIZE / 2, yCursor);
-      this.p5.textSize(FONT_SIZE);
+    let renderInfo = {
+      color: poemState.color,
+      font: poemState.font,
+      lines: []
+    };
+    if (poemState.title) {
+      renderInfo.title = {
```

Comment:
Not sure if it will change in a future PR, but with the renderInfo for title/author/lines, it seems like there is some duplicated information that we end up passing through the renderInfo, like the X value (and the pattern of calculating the Y value), is it worth pulling out some of the duplication to make it easier to read?

Reply:
Sorry, not sure I'm 100% following..Can you explain a little more what should be pulled out?